src/parseTree.js

```diff
@@ -281,37 +270,121 @@
     props: {
       traceNode: {type: Object, required: true},
       measureInputText: {type: Function},
-      classes: {type: Object, default: Object},
       isInVBox: {type: Boolean},
-      vbox: {type: Boolean}
+
+      // from parent element
+      context: {type: Object},
+
+      // from currentLR
+      currentLR: {type: Object, default: Object},
+      eventHandlers: {type: Object}
     },
     computed: {
+      labeled: function() {
+        return shouldNodeBeLabeled(this.traceNode);
+      },
+      vbox: function() {
+        return hasVisibleChoice(this.traceNode) ||
+               hasVisibleLeftRecursion(this.traceNode) ||
+               (isAlt(this.traceNode.expr) && this.context && this.context.vbox);
+      },
+      isWhitespace: function() {
+        return this.traceNode.ruleName === 'spaces';
+      },
+      isLeaf: function() {
+        var leaf = isLeaf(this.traceNode);
+        if (this.traceNode.isMemoized) {
+          var memoKey = this.traceNode.expr.toMemoKey();
+          var stack = this.currentLR[memoKey];
+          if (stack && stack[stack.length - 1] === this.traceNode.pos) {
+            leaf = true;
+          }
+        }
+        return leaf;
+      },
       classObj: function() {
         var obj = {
-          disclosure: this.classes.labeled && this.isInVBox
+          disclosure: this.labeled && this.isInVBox
         };
         var ctorName = this.traceNode.ctorName;
         if (ctorName) {
           obj[ctorName.toLowerCase()] = true;
         }
-        Object.assign(obj, this.classes);
+
+        obj.collapsed = this.labeled &&
```

Comment:
I rewrited the whole computed children part. As you suggested, I add a `collapsed`  into data for recompute `children`.  However, I am keeping the `collapsed` in the class object as well. Here are my thoughts:

`this.collapsed` is used for re-rendering children only, and it only changed from `true` to `false` once when a node is uncollapsed. I am following the logic that the children only get rendered the first time its parent get uncollapsed. That means after the children rendered, it'll always be there for the same tree. So after a node's children get rendered, toggle it will only show/hide the existed children correspondingly without any re-rendering. I think rendering children every time a node get uncollapsed could hurt the performance. Correct me if I am wrong, if it's not the case, I'll write it in the way exactly as you suggested.

`this.classObj.collapsed` is for styling the dom node in general, which the real collapsed status for the node.

That's the reason I keep both of them. Hope it's clear, and let me know if you have any comment.